### ISSUE TITLE
NoSuchMethodError when trying to set display name and lore on ItemMetaMock

### ISSUE THREAD

**Issue by natereprogle** [MockBukkit/MockBukkit#462]
I'm receiving a NoSuchMethod error when trying to test ItemMocks. It appears that on line 75 of ItemMetaMock.java, the following method is used:

```java
if (meta.hasDisplayName())
{
    displayName = GsonComponentSerializer.gson().serialize(meta.displayName());
}
```

However, ItemMeta doesn't have a `#displayName()` function, at least in 1.19-R0.1-SNAPSHOT. It has a `#getDisplayName()` and `#setDisplayName()` function. It appears the same thing is happening with ItemMeta lore, because the `#lore()` function is being used instead of `#getLore()`

```
java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
 at be.seeseemelk.mockbukkit.inventory.ItemFactoryMock.asMetaFor([ItemFactoryMock.java:143](command:_java.test.openStackTrace?%5B%22%5Ctat%20be.seeseemelk.mockbukkit.inventory.ItemFactoryMock.asMetaFor(ItemFactoryMock.java%3A143)%22%2C%22honeypot-core%22%5D))
 at org.bukkit.inventory.ItemStack.setItemMeta0([ItemStack.java:585](command:_java.test.openStackTrace?%5B%22%5Ctat%20org.bukkit.inventory.ItemStack.setItemMeta0(ItemStack.java%3A585)%22%2C%22honeypot-core%22%5D))
 at org.bukkit.inventory.ItemStack.setItemMeta([ItemStack.java:571](command:_java.test.openStackTrace?%5B%22%5Ctat%20org.bukkit.inventory.ItemStack.setItemMeta(ItemStack.java%3A571)%22%2C%22honeypot-core%22%5D))
 at org.reprogle.honeypot.gui.TestGuiButtonCreation.createDummyitem([TestGuiButtonCreation.java:102](command:_java.test.openStackTrace?%5B%22%5Ctat%20org.reprogle.honeypot.gui.TestGuiButtonCreation.createDummyitem(TestGuiButtonCreation.java%3A102)%22%2C%22honeypot-core%22%5D))
 at org.reprogle.honeypot.gui.TestGuiButtonCreation.testButtons([TestGuiButtonCreation.java:84](command:_java.test.openStackTrace?%5B%22%5Ctat%20org.reprogle.honeypot.gui.TestGuiButtonCreation.testButtons(TestGuiButtonCreation.java%3A84)%22%2C%22honeypot-core%22%5D))
 at java.base/java.util.ArrayList.forEach([ArrayList.java:1511](command:_java.test.openStackTrace?%5B%22%5Ctat%20java.base%2Fjava.util.ArrayList.forEach(ArrayList.java%3A1511)%22%2C%22honeypot-core%22%5D))
 at java.base/java.util.ArrayList.forEach([ArrayList.java:1511](command:_java.test.openStackTrace?%5B%22%5Ctat%20java.base%2Fjava.util.ArrayList.forEach(ArrayList.java%3A1511)%22%2C%22honeypot-core%22%5D))
Caused by: java.lang.reflect.InvocationTargetException
 at java.base/java.lang.reflect.Constructor.newInstanceWithCaller([Constructor.java:499](command:_java.test.openStackTrace?%5B%22%5Ctat%20java.base%2Fjava.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java%3A499)%22%2C%22honeypot-core%22%5D))
 at java.base/java.lang.reflect.Constructor.newInstance([Constructor.java:480](command:_java.test.openStackTrace?%5B%22%5Ctat%20java.base%2Fjava.lang.reflect.Constructor.newInstance(Constructor.java%3A480)%22%2C%22honeypot-core%22%5D))
 at be.seeseemelk.mockbukkit.inventory.ItemFactoryMock.asMetaFor([ItemFactoryMock.java:133](command:_java.test.openStackTrace?%5B%22%5Ctat%20be.seeseemelk.mockbukkit.inventory.ItemFactoryMock.asMetaFor(ItemFactoryMock.java%3A133)%22%2C%22honeypot-core%22%5D))
 ... 73 more
Caused by: java.lang.NoSuchMethodError: 'net.kyori.adventure.text.Component org.bukkit.inventory.meta.ItemMeta.displayName()'
 at be.seeseemelk.mockbukkit.inventory.meta.ItemMetaMock.([ItemMetaMock.java:75](command:_java.test.openStackTrace?%5B%22%5Ctat%20be.seeseemelk.mockbukkit.inventory.meta.ItemMetaMock.%3Cinit%3E(ItemMetaMock.java%3A75)%22%2C%22honeypot-core%22%5D))
 ... 79 more
```

**Comment by Jikoo** [comment]
Having a similar issue, only with ServerMock#addPlayer. MockBukkit is no longer capable of testing Bukkit plugins after the Paper changes. If MockBukkit really wants to go the Paper route I think it should clarify that (both the MockBukkit name and the readme indicate a commitment to Bukkit/CB consistency) or ensure it is loading Paper classes at test runtime.

The "easy" solution is to mandate that users' projects are compiled with Gradle. That way they can compile against Bukkit and test against Paper.

I personally am just going to begin phasing MockBukkit out of my projects as it fails instead. Mockito may be slower, but it's consistent. I'd rather have my tests take a little longer to execute and just work than continue investing time in using a resource that requires me to jump through a new hoop every few updates.

**Comment by thelooter** [comment]
You can still use the Versions before 2.0.0, they still work with Spigot

**Comment by thelooter** [comment]
Also, what prohibits you from switching to PaperMC? We're currently discussing how to handle this in the Core Team, so any amount of Information would be helpful

**Comment by natereprogle** [comment]
> Also, what prohibits you from switching to PaperMC? We're currently discussing how to handle this in the Core Team, so any amount of Information would be helpful

Ahh, I didn't know MockBukkit required Paper, I thought it supported Spigot as well. I use Spigot for the quicker update cycles and larger range of supported servers that can potentially use my plugin. They had 1.19 support out on day 1. Not all servers will use Paper, but guaranteed most servers will run Spigot. This is just a design choice and there is no _real_ reason stopping me from switching.

**Comment by natereprogle** [comment]
> Having a similar issue, only with ServerMock#addPlayer. MockBukkit is no longer capable of testing Bukkit plugins after the Paper changes. If MockBukkit really wants to go the Paper route I think it should clarify that (both the MockBukkit name and the readme indicate a commitment to Bukkit/CB consistency) or ensure it is loading Paper classes at test runtime.
> 
> 
> 
> The "easy" solution is to mandate that users' projects are compiled with Gradle. That way they can compile against Bukkit and test against Paper.
> 
> 
> 
> I personally am just going to begin phasing MockBukkit out of my projects as it fails instead. Mockito may be slower, but it's consistent. I'd rather have my tests take a little longer to execute and just work than continue investing time in using a resource that requires me to jump through a new hoop every few updates.

I was going to open an issue on this as well, but figured it could be discussed here. I'm slowly learning about unit testing, so I'm not familiar with all the libraries out there and their capabilities.

**Comment by Jikoo** [comment]
> You can still use the Versions before 2.0.0, they still work with Spigot

Sticking with 1.X defeats the whole purpose of MockBukkit. If I have to implement everything new myself, I may as well just use Mockito from the start.

> Also, what prohibits you from switching to PaperMC? We're currently discussing how to handle this in the Core Team, so any amount of Information would be helpful

Spigot still has a 31% server market share. Another 2% or so is various forks that likely are Bukkit API implementations. While I agree that Paper's server software is objectively superior in every way and people should swap, supporting Bukkit nets me 33% more downloads, and downloads are ego pats. Now that I think about it though, my number-based reasoning is horribly illogical, because I also generally do not support versions beyond the last 2 MC minor versions.

Pretty much all my unpublished work is for Paper, but pretty much all my published work is for Bukkit so that anyone who finds it can use it. I largely don't need Paper's API changes - there are some convenience methods, a few useful events, but by and large unless I am working with text or entity AI Bukkit is sufficient. Compiling against Bukkit means that I don't risk accidentally using a Paper convenience method.

I think the best solution is to have Mockbukkit encourage adopting Gradle - it really is a much nicer build tool. That way we Bukkit dinosaurs can still have our numbers, but people using Paper can have thorough tests.

**Comment by thelooter** [comment]
I see, didn't think about the Downloads Part. The reason Spigot has Day 1 Releases is that they can work of Snapshots while Paper has to wait for the Spigot work to be released. We're currently discussing ways to support multiple APIs and Minecraft Versions, so if you are interested in joining that Discussion, feel free to join the Discord Server and message me.

**Comment by thelooter** [comment]
> The "easy" solution is to mandate that users' projects are compiled with Gradle. That way they can compile against Bukkit and test against Paper.

Isn't that a possibility with the Maven `test` Scope as well

**Comment by natereprogle** [comment]
> > The "easy" solution is to mandate that users' projects are compiled with Gradle. That way they can compile against Bukkit and test against Paper.
> 
> Isn't that a possibility with the Maven `test` Scope as well

I was going to pop in and reply as well, I assume you can use Maven's `test` to test against Paper and Build against spigot. I'm going to try that and see what happens, I'll let you know.

**Comment by Jikoo** [comment]
> > The "easy" solution is to mandate that users' projects are compiled with Gradle. That way they can compile against Bukkit and test against Paper.
> 
> Isn't that a possibility with the Maven `test` Scope as well

My initial response was no - Maven does not have a `testCompile` scope like Gradle - but on further testing it turns out that Maven does not resolve the entire classpath until the corresponding scope, so yes, but only if MockBukkit is declared prior to Bukkit in the dependency list so that its transitive Paper dependency overrides Bukkit. That didn't occur to me because it's the opposite of standard practice, test dependencies tend to be last. Turns out I'm making a fuss about nothing, I'm sorry for wasting your time.

Tl;dr: for me and OP: MockBukkit needs to be declared first now.
```xml
<dependencies>
  <dependency>
    <groupId>com.github.seeseemelk</groupId>
    <artifactId>MockBukkit-v1.19</artifactId>
    <version>2.38.0</version>
    <scope>test</scope>
  </dependency>
  <dependency>
    <groupId>org.spigotmc</groupId>
    <artifactId>spigot-api</artifactId>
    <version>1.19-R0.1-SNAPSHOT</version>
    <scope>provided</scope>
  </dependency>
  ...
```

**Comment by thelooter** [comment]
This still is a good point you're bringing up, so we'll have some further discussions in the Core Team and hopefully come up with a less hacky solution

**Comment by natereprogle** [comment]
> > > The "easy" solution is to mandate that users' projects are compiled with Gradle. That way they can compile against Bukkit and test against Paper.
> 
> > 
> 
> > Isn't that a possibility with the Maven `test` Scope as well
> 
> 
> 
> I was going to pop in and reply as well, I assume you can use Maven's `test` to test against Paper and Build against spigot. I'm going to try that and see what happens, I'll let you know.

I didn't test adding a player as Jikoo suggested, but I did test with my ItemMeta mocks. Using Paper for testing worked, so thank you for that suggestion! I was unaware Paper was required for using MockBukkit which is why I incorrectly filed this Issue. If, at some point in the future, MockBukkit can support Spigot/Bukkit that would be nice, but imo it's not really necessary since there is a really easy "workaround" 😃

**Comment by thelooter** [comment]
Yeah, since 2.0.0 we build on Paper. And since paper internally handles a lot of stuff with @KyoriPowered Components, it will fail on Spigot

**Comment by thelooter** [comment]
We decided on a Method to fix this issue, but it is a huge effort so It might take some time

**Comment by Insprill** [comment]
We've decided not to proceed with our original idea due to the additional maintenance overhead. MockBukkit now officially only supports Paper, though we'll try to keep Spigot support where possible. One possible solution would be adding Paper as a test-only dependency.
